src/main/java/graphql/execution/ExecutionContext.java

```diff
@@ -80,7 +84,17 @@ public class ExecutionContext {
         this.errors.set(builder.errors);
         this.localContext = builder.localContext;
         this.executionInput = builder.executionInput;
-        queryTree = FpKit.interThreadMemoize(() -> ExecutableNormalizedOperationFactory.createExecutableNormalizedOperation(graphQLSchema, operationDefinition, fragmentsByName, coercedVariables));
+        this.queryTree = FpKit.interThreadMemoize(() -> ExecutableNormalizedOperationFactory.createExecutableNormalizedOperation(graphQLSchema, operationDefinition, fragmentsByName, coercedVariables));
+        this.rootDataFetchingFieldSelectionSet = FpKit.intraThreadMemoize(() -> collectRootDataFetchingFieldSelectionSets(queryTree, builder.graphQLSchema));
+    }
+
+    @NotNull
+    private List<DataFetchingFieldSelectionSet> collectRootDataFetchingFieldSelectionSets(
+            Supplier<ExecutableNormalizedOperation> queryTree, GraphQLSchema schema) {
+        return ImmutableKit.map(queryTree.get().getTopLevelFields(), root -> DataFetchingFieldSelectionSetImpl.newCollector(
+                schema,
+                root.getFieldDefinitions(schema).get(0).getType(), //root fields always have a single parent
```

Comment:
This is, I believe, safe. All root fields have a singular parent, introspection fields included. Right?